File: python/forbid.py
#!/usr/bin/env python
# _*_ encoding: utf-8 _*_

import sys
import re
import json
import time

domain_forbid = {}
app_forbid = {}
stream_forbid = {}

def to_str(*val):
    ans = ""
    split = ""
    for v in val:
        ans = ans + split + str(v)
        split = ":"

    return str(ans)

def to_number(val):
    return int(val)

def forbid_filter(dicts, key, endtime):

    if key in dicts:
        if endtime == 0:
            dicts.pop(key)
        elif dicts[key] < endtime:
            dicts[key] = endtime
    else:
        cur_time = int(time.time())
        if endtime != 0 and endtime > cur_time:
            dicts[key] = endtime


def write_forbid_file(name, data):

    with open(name, 'w') as fp:
        json.dump(data, fp, indent=4)


if __name__ == '__main__':

    for line in open(sys.argv[1], 'r'):
        if re.findall(r'/stream_forbid/domain', line):
            domain = re.findall(r'domain=([^&]*)', line)
            end_time = re.findall(r'end_time=(\d*)', line)
            if end_time and domain:
                dm = domain[0]
                et = end_time[0]
                forbid_filter(domain_forbid, to_str(dm), to_number(et))
        elif re.findall(r'/stream_forbid/app', line):
            domain = re.findall(r'domain=([^&]*)', line)
            app = re.findall(r'app=([^&]*)', line)
            end_time = re.findall(r'end_time=(\d*)', line)
            if end_time and domain and app:
                dm = domain[0]
                ap = app[0]
                et = end_time[0]
                forbid_filter(app_forbid, to_str(dm, ap), to_number(et))
        elif re.findall(r'/stream_forbid/stream', line):
            domain = re.findall(r'domain=([^&]*)', line)
            app = re.findall(r'app=([^&]*)', line)
            stream = re.findall(r'stream=([^&]*)', line)
            end_time = re.findall(r'end_time=(\d*)', line)
            if end_time and domain and app and stream:
                dm = domain[0]
                ap = app[0]
                sm = stream[0]
                et = end_time[0]
                forbid_filter(stream_forbid, to_str(dm, ap, sm), to_number(et))
        else:
            print(line)


    write_forbid_file("domain_forbid.json", domain_forbid)
    write_forbid_file("app_forbid.json", app_forbid)
    write_forbid_file("stream_forbid.json", stream_forbid)

File: perl/forbid.pl
#!/usr/bin/env perl

use strict;
use warnings;
use utf8;

# cpan JSON
use JSON;

my %forbid_record;

my $file = $ARGV[0];

sub say {print @_, "\n"}

sub to_str {

    my $ans = "";
    my $split = "";
    foreach (@_) {
        $ans .= $split .= $_;
        $split = ":";
    }

    return $ans;
}

sub filter_forbid {
    my ($key, $endtime) = @_;

    if ( exists($forbid_record{$key}) ) {
        if ($endtime == 0) {
            delete $forbid_record{$key};
        } elsif ($forbid_record{$key} < $endtime) {
            $forbid_record{$key} = $endtime;
        }
    } else {
        my $cur_time = time();
        if ($endtime != 0 and $endtime > $cur_time ) {
            $forbid_record{$key} = $endtime;
        }
    }
}

sub dump_json_file {
    my ($file) = @_;

    open my $fh, ">", $file or die "Counld open $file: $!";
    print $fh encode_json \%forbid_record;
    close $fh;
}

open my $info, $file or die "Could not open $file: $!";

while( my $line = <$info> ) {

    my $domain = "";
    my $app = "all";
    my $stream = "all";
    my $end_time = "";

    if ($line =~ "/stream_forbid/domain") {
        # filter domain
        if ($line =~ "domain=(?<domain>[^&]*)") {
            $domain = $+{domain};
            # filter end_time
            if ($line =~ "end_time=(?<end_time>\\d+)") {
                $end_time = $+{end_time};
                # handle
                filter_forbid(to_str($domain, $app, $stream), $end_time);
            }
        }
    } elsif($line =~ "/stream_forbid/app") {
        # filter domain
        if ($line =~ "domain=(?<domain>[^&]*)") {
            $domain = $+{domain};
            # filter app
            if ($line =~ "app=(?<app>[^&]*)") {
                $app = $+{app};
                # filter end_time
                if ($line =~ "end_time=(?<end_time>\\d+)") {
                    $end_time = $+{end_time};
                    # handle
                    filter_forbid(to_str($domain, $app, $stream), $end_time);
                }
            }
        }
    } elsif($line =~ "/stream_forbid/stream") {
        # filter domain
        if ($line =~ "domain=(?<domain>[^&]*)") {
            $domain = $+{domain};
            # filter app
            if ($line =~ "app=(?<app>[^&]*)") {
                $app = $+{app};
                # filter stream
                if ($line =~ "stream=(?<stream>[^&]*)") {
                    $stream = $+{stream};
                    # filter end_time
                    if ($line =~ "end_time=(?<end_time>\\d+)") {
                        $end_time = $+{end_time};
                        # handle
                        filter_forbid(to_str($domain, $app, $stream), $end_time);
                    }
                }
            }
        }
    } else {
        say $line
    }
}

dump_json_file("/tmp/forbid.json");

close $info;
